{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>crassus_id</th>\n",
       "      <th>contig</th>\n",
       "      <th>sample</th>\n",
       "      <th>length</th>\n",
       "      <th>length/ref</th>\n",
       "      <th>ref</th>\n",
       "      <th>DTR</th>\n",
       "      <th>coding</th>\n",
       "      <th>family</th>\n",
       "      <th>subfamily</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>evidence_family</th>\n",
       "      <th>evidence_genus</th>\n",
       "      <th>notes</th>\n",
       "      <th>discard</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IBD-crossAssembled_1_182976</td>\n",
       "      <td>contig_241570</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>182976</td>\n",
       "      <td>0.978</td>\n",
       "      <td>Zeta_2</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>Zeta</td>\n",
       "      <td>Zeta</td>\n",
       "      <td>Zeta_2</td>\n",
       "      <td>species__72</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>IBD-crossAssembled_2_178222</td>\n",
       "      <td>contig_8225</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>178222</td>\n",
       "      <td>0.975</td>\n",
       "      <td>Zeta_3_1</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Zeta</td>\n",
       "      <td>Zeta</td>\n",
       "      <td>Zeta_3</td>\n",
       "      <td>Zeta_3_1</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots, phylogenies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IBD-crossAssembled_4_145297</td>\n",
       "      <td>contig_1091</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>145297</td>\n",
       "      <td>0.954</td>\n",
       "      <td>Epsilon_2_3</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Epsilon</td>\n",
       "      <td>Epsilon</td>\n",
       "      <td>Epsilon_2</td>\n",
       "      <td>Epsilon_2_3</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots, phylogenies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IBD-crossAssembled_7_141010</td>\n",
       "      <td>contig_1857</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>141010</td>\n",
       "      <td>0.922</td>\n",
       "      <td>Epsilon_1</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Epsilon</td>\n",
       "      <td>Epsilon</td>\n",
       "      <td>Epsilon_1</td>\n",
       "      <td>species__92</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots, phylogenies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IBD-crossAssembled_16_113452</td>\n",
       "      <td>contig_17571</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>113452</td>\n",
       "      <td>0.640</td>\n",
       "      <td>Zeta_7_1</td>\n",
       "      <td>False</td>\n",
       "      <td>TGA</td>\n",
       "      <td>Zeta</td>\n",
       "      <td>Zeta</td>\n",
       "      <td>Zeta_7</td>\n",
       "      <td>Zeta_7_1</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots, phylogenies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>IBD-crossAssembled_3122_11376</td>\n",
       "      <td>contig_7298</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>11376</td>\n",
       "      <td>0.118</td>\n",
       "      <td>Buorbuivirus</td>\n",
       "      <td>False</td>\n",
       "      <td>TAG</td>\n",
       "      <td>Suoliviridae</td>\n",
       "      <td>Loutivirinae</td>\n",
       "      <td>Buorbuivirus</td>\n",
       "      <td>species__173</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>IBD-crossAssembled_3177_11243</td>\n",
       "      <td>contig_20815</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>11243</td>\n",
       "      <td>0.113</td>\n",
       "      <td>Kolpuevirus hominis</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "      <td>Asinivirinae</td>\n",
       "      <td>Kolpuevirus</td>\n",
       "      <td>Kolpuevirus hominis</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots, phylogenies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>IBD-crossAssembled_3515_10591</td>\n",
       "      <td>contig_239060</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>10591</td>\n",
       "      <td>0.109</td>\n",
       "      <td>Carjivirus communis</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Intestiviridae</td>\n",
       "      <td>Crudevirinae</td>\n",
       "      <td>Carjivirus</td>\n",
       "      <td>Carjivirus communis</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots, phylogenies</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>IBD-crossAssembled_3720_10264</td>\n",
       "      <td>contig_240585</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>10264</td>\n",
       "      <td>0.112</td>\n",
       "      <td>Dabirmavirus</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Intestiviridae</td>\n",
       "      <td>Crudevirinae</td>\n",
       "      <td>Dabirmavirus</td>\n",
       "      <td>species__174</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>IBD-crossAssembled_3833_10082</td>\n",
       "      <td>contig_241434</td>\n",
       "      <td>IBD-crossAssembled</td>\n",
       "      <td>10082</td>\n",
       "      <td>0.101</td>\n",
       "      <td>Kolpuevirus hominis</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "      <td>Asinivirinae</td>\n",
       "      <td>Kolpuevirus</td>\n",
       "      <td>Kolpuevirus hominis</td>\n",
       "      <td>phylogenies, shared_prots</td>\n",
       "      <td>ani, shared_prots</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        crassus_id         contig              sample  length  \\\n",
       "0      IBD-crossAssembled_1_182976  contig_241570  IBD-crossAssembled  182976   \n",
       "1      IBD-crossAssembled_2_178222    contig_8225  IBD-crossAssembled  178222   \n",
       "2      IBD-crossAssembled_4_145297    contig_1091  IBD-crossAssembled  145297   \n",
       "3      IBD-crossAssembled_7_141010    contig_1857  IBD-crossAssembled  141010   \n",
       "4     IBD-crossAssembled_16_113452   contig_17571  IBD-crossAssembled  113452   \n",
       "..                             ...            ...                 ...     ...   \n",
       "106  IBD-crossAssembled_3122_11376    contig_7298  IBD-crossAssembled   11376   \n",
       "107  IBD-crossAssembled_3177_11243   contig_20815  IBD-crossAssembled   11243   \n",
       "108  IBD-crossAssembled_3515_10591  contig_239060  IBD-crossAssembled   10591   \n",
       "109  IBD-crossAssembled_3720_10264  contig_240585  IBD-crossAssembled   10264   \n",
       "110  IBD-crossAssembled_3833_10082  contig_241434  IBD-crossAssembled   10082   \n",
       "\n",
       "     length/ref                  ref    DTR coding          family  \\\n",
       "0         0.978               Zeta_2   True     11            Zeta   \n",
       "1         0.975             Zeta_3_1  False     11            Zeta   \n",
       "2         0.954          Epsilon_2_3  False     11         Epsilon   \n",
       "3         0.922            Epsilon_1  False     11         Epsilon   \n",
       "4         0.640             Zeta_7_1  False    TGA            Zeta   \n",
       "..          ...                  ...    ...    ...             ...   \n",
       "106       0.118         Buorbuivirus  False    TAG    Suoliviridae   \n",
       "107       0.113  Kolpuevirus hominis  False     11    Steigviridae   \n",
       "108       0.109  Carjivirus communis  False     11  Intestiviridae   \n",
       "109       0.112         Dabirmavirus  False     11  Intestiviridae   \n",
       "110       0.101  Kolpuevirus hominis  False     11    Steigviridae   \n",
       "\n",
       "        subfamily         genus              species  \\\n",
       "0            Zeta        Zeta_2          species__72   \n",
       "1            Zeta        Zeta_3             Zeta_3_1   \n",
       "2         Epsilon     Epsilon_2          Epsilon_2_3   \n",
       "3         Epsilon     Epsilon_1          species__92   \n",
       "4            Zeta        Zeta_7             Zeta_7_1   \n",
       "..            ...           ...                  ...   \n",
       "106  Loutivirinae  Buorbuivirus         species__173   \n",
       "107  Asinivirinae   Kolpuevirus  Kolpuevirus hominis   \n",
       "108  Crudevirinae    Carjivirus  Carjivirus communis   \n",
       "109  Crudevirinae  Dabirmavirus         species__174   \n",
       "110  Asinivirinae   Kolpuevirus  Kolpuevirus hominis   \n",
       "\n",
       "               evidence_family                  evidence_genus notes  discard  \n",
       "0    phylogenies, shared_prots               ani, shared_prots   NaN    False  \n",
       "1    phylogenies, shared_prots  ani, shared_prots, phylogenies   NaN    False  \n",
       "2    phylogenies, shared_prots  ani, shared_prots, phylogenies   NaN    False  \n",
       "3    phylogenies, shared_prots  ani, shared_prots, phylogenies   NaN    False  \n",
       "4    phylogenies, shared_prots  ani, shared_prots, phylogenies   NaN    False  \n",
       "..                         ...                             ...   ...      ...  \n",
       "106  phylogenies, shared_prots                             ani   NaN    False  \n",
       "107  phylogenies, shared_prots  ani, shared_prots, phylogenies   NaN    False  \n",
       "108  phylogenies, shared_prots  ani, shared_prots, phylogenies   NaN    False  \n",
       "109  phylogenies, shared_prots                             ani   NaN    False  \n",
       "110  phylogenies, shared_prots               ani, shared_prots   NaN    False  \n",
       "\n",
       "[111 rows x 16 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crassus = pd.read_csv(\"crassus_results.tsv.gz\", delimiter=\"\\t\", compression='gzip')\n",
    "crassus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "crassshort = crassus[['contig', 'length', 'ref', 'coding', 'family']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>contig</th>\n",
       "      <th>length</th>\n",
       "      <th>Reads</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>SPM</th>\n",
       "      <th>averageFold</th>\n",
       "      <th>referenceGC</th>\n",
       "      <th>coveragePercentage</th>\n",
       "      <th>coverageBases</th>\n",
       "      <th>medianFold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_10493</td>\n",
       "      <td>18937</td>\n",
       "      <td>2</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.3441</td>\n",
       "      <td>4.18</td>\n",
       "      <td>792</td>\n",
       "      <td>0.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_1091</td>\n",
       "      <td>145297</td>\n",
       "      <td>3</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.3817</td>\n",
       "      <td>0.50</td>\n",
       "      <td>724</td>\n",
       "      <td>0.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_12034</td>\n",
       "      <td>21580</td>\n",
       "      <td>485</td>\n",
       "      <td>11.80</td>\n",
       "      <td>11.78</td>\n",
       "      <td>57.58</td>\n",
       "      <td>9.03</td>\n",
       "      <td>0.3575</td>\n",
       "      <td>85.35</td>\n",
       "      <td>18418</td>\n",
       "      <td>8.32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_12041</td>\n",
       "      <td>29559</td>\n",
       "      <td>266</td>\n",
       "      <td>4.73</td>\n",
       "      <td>4.73</td>\n",
       "      <td>20.76</td>\n",
       "      <td>3.25</td>\n",
       "      <td>0.3691</td>\n",
       "      <td>66.94</td>\n",
       "      <td>19786</td>\n",
       "      <td>4.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_12042</td>\n",
       "      <td>11863</td>\n",
       "      <td>108</td>\n",
       "      <td>4.78</td>\n",
       "      <td>4.78</td>\n",
       "      <td>20.00</td>\n",
       "      <td>3.13</td>\n",
       "      <td>0.3659</td>\n",
       "      <td>64.55</td>\n",
       "      <td>7658</td>\n",
       "      <td>3.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18474</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_5776</td>\n",
       "      <td>88456</td>\n",
       "      <td>6</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.3260</td>\n",
       "      <td>0.96</td>\n",
       "      <td>853</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18475</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_7024</td>\n",
       "      <td>36085</td>\n",
       "      <td>6</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.3279</td>\n",
       "      <td>3.68</td>\n",
       "      <td>1328</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18476</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_7034</td>\n",
       "      <td>16937</td>\n",
       "      <td>1</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.3478</td>\n",
       "      <td>1.38</td>\n",
       "      <td>234</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18477</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_7054</td>\n",
       "      <td>34118</td>\n",
       "      <td>4</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.3275</td>\n",
       "      <td>2.98</td>\n",
       "      <td>1016</td>\n",
       "      <td>0.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18478</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_7091</td>\n",
       "      <td>66998</td>\n",
       "      <td>8</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.3245</td>\n",
       "      <td>2.73</td>\n",
       "      <td>1832</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18479 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample        contig  length  Reads   RPKM   FPKM    SPM  averageFold  \\\n",
       "0        7544  contig_10493   18937      2   0.06   0.06   0.35         0.05   \n",
       "1        7544   contig_1091  145297      3   0.01   0.01   0.03         0.01   \n",
       "2        7544  contig_12034   21580    485  11.80  11.78  57.58         9.03   \n",
       "3        7544  contig_12041   29559    266   4.73   4.73  20.76         3.25   \n",
       "4        7544  contig_12042   11863    108   4.78   4.78  20.00         3.13   \n",
       "...       ...           ...     ...    ...    ...    ...    ...          ...   \n",
       "18474   38211   contig_5776   88456      6   0.03   0.03   0.11         0.02   \n",
       "18475   38211   contig_7024   36085      6   0.08   0.08   0.22         0.04   \n",
       "18476   38211   contig_7034   16937      1   0.03   0.00   0.07         0.01   \n",
       "18477   38211   contig_7054   34118      4   0.06   0.06   0.19         0.04   \n",
       "18478   38211   contig_7091   66998      8   0.06   0.06   0.18         0.04   \n",
       "\n",
       "       referenceGC  coveragePercentage  coverageBases  medianFold  \n",
       "0           0.3441                4.18            792        0.28  \n",
       "1           0.3817                0.50            724        0.07  \n",
       "2           0.3575               85.35          18418        8.32  \n",
       "3           0.3691               66.94          19786        4.40  \n",
       "4           0.3659               64.55           7658        3.64  \n",
       "...            ...                 ...            ...         ...  \n",
       "18474       0.3260                0.96            853        0.23  \n",
       "18475       0.3279                3.68           1328        0.23  \n",
       "18476       0.3478                1.38            234        0.12  \n",
       "18477       0.3275                2.98           1016        0.22  \n",
       "18478       0.3245                2.73           1832        0.23  \n",
       "\n",
       "[18479 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contigs = pd.read_csv(\"crass_contigs.tsv.gz\", delimiter=\"\\t\", compression='gzip')\n",
    "contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{33855,\n",
       " 33856,\n",
       " 33857,\n",
       " 33858,\n",
       " 33859,\n",
       " 33860,\n",
       " 33861,\n",
       " 33862,\n",
       " 33863,\n",
       " 33865,\n",
       " 33866,\n",
       " 33867,\n",
       " 33868,\n",
       " 33869,\n",
       " 33870,\n",
       " 33871,\n",
       " 33872,\n",
       " 33877,\n",
       " 33878,\n",
       " 33879,\n",
       " 33880,\n",
       " 33881,\n",
       " 33882,\n",
       " 33883,\n",
       " 33884,\n",
       " 33887,\n",
       " 33888,\n",
       " 33890,\n",
       " 33891,\n",
       " 33893,\n",
       " 33895,\n",
       " 33896,\n",
       " 33897,\n",
       " 33898,\n",
       " 33901,\n",
       " 33902,\n",
       " 33906,\n",
       " 33907,\n",
       " 33908,\n",
       " 33909,\n",
       " 33910,\n",
       " 33911,\n",
       " 33913,\n",
       " 33914,\n",
       " 33915,\n",
       " 33916,\n",
       " 33917,\n",
       " 33922,\n",
       " 33924,\n",
       " 33925,\n",
       " 33926,\n",
       " 33929,\n",
       " 33931,\n",
       " 33932,\n",
       " 33936,\n",
       " 33937,\n",
       " 33939,\n",
       " 33940,\n",
       " 33941,\n",
       " 33942,\n",
       " 33943,\n",
       " 33947,\n",
       " 33949,\n",
       " 33950,\n",
       " 33951,\n",
       " 33953,\n",
       " 33954,\n",
       " 33955,\n",
       " 33957,\n",
       " 33958,\n",
       " 33959,\n",
       " 33960,\n",
       " 33961,\n",
       " 33963,\n",
       " 33964,\n",
       " 33965,\n",
       " 33966,\n",
       " 33968,\n",
       " 33969,\n",
       " 33970,\n",
       " 33972,\n",
       " 33973,\n",
       " 33974,\n",
       " 33975,\n",
       " 33976,\n",
       " 33978,\n",
       " 33979,\n",
       " 33980,\n",
       " 33981,\n",
       " 33982,\n",
       " 33983,\n",
       " 33984,\n",
       " 33988,\n",
       " 33989,\n",
       " 33991,\n",
       " 33992,\n",
       " 33993,\n",
       " 33995,\n",
       " 33996,\n",
       " 33997,\n",
       " 33998,\n",
       " 33999,\n",
       " 34000,\n",
       " 34001,\n",
       " 34002,\n",
       " 34003,\n",
       " 34006,\n",
       " 34007,\n",
       " 34008,\n",
       " 34009,\n",
       " 34010,\n",
       " 34011,\n",
       " 34012,\n",
       " 34013,\n",
       " 34015,\n",
       " 34018,\n",
       " 34019,\n",
       " 34020,\n",
       " 34022,\n",
       " 34024,\n",
       " 34027,\n",
       " 34028,\n",
       " 34029,\n",
       " 34030,\n",
       " 34031,\n",
       " 34033,\n",
       " 34035,\n",
       " 34036,\n",
       " 34037,\n",
       " 34038,\n",
       " 34039,\n",
       " 34040,\n",
       " 34041,\n",
       " 34042,\n",
       " 34043,\n",
       " 34044,\n",
       " 34054}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wehave = set()\n",
    "with open(\"freezer.txt\", 'r') as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if l:\n",
    "            wehave.add(int(l))\n",
    "wehave"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>contig</th>\n",
       "      <th>length</th>\n",
       "      <th>Reads</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>SPM</th>\n",
       "      <th>averageFold</th>\n",
       "      <th>referenceGC</th>\n",
       "      <th>coveragePercentage</th>\n",
       "      <th>coverageBases</th>\n",
       "      <th>medianFold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6294</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>14547</td>\n",
       "      <td>1228</td>\n",
       "      <td>70.77</td>\n",
       "      <td>70.77</td>\n",
       "      <td>173.98</td>\n",
       "      <td>19.77</td>\n",
       "      <td>0.3652</td>\n",
       "      <td>88.46</td>\n",
       "      <td>12868</td>\n",
       "      <td>23.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6295</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_11000</td>\n",
       "      <td>18137</td>\n",
       "      <td>350</td>\n",
       "      <td>16.18</td>\n",
       "      <td>16.18</td>\n",
       "      <td>38.78</td>\n",
       "      <td>4.41</td>\n",
       "      <td>0.3337</td>\n",
       "      <td>48.42</td>\n",
       "      <td>8782</td>\n",
       "      <td>7.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6296</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_11002</td>\n",
       "      <td>106707</td>\n",
       "      <td>894</td>\n",
       "      <td>7.02</td>\n",
       "      <td>7.02</td>\n",
       "      <td>17.31</td>\n",
       "      <td>1.97</td>\n",
       "      <td>0.3259</td>\n",
       "      <td>23.40</td>\n",
       "      <td>24971</td>\n",
       "      <td>6.93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6297</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_11067</td>\n",
       "      <td>24409</td>\n",
       "      <td>72</td>\n",
       "      <td>2.47</td>\n",
       "      <td>2.47</td>\n",
       "      <td>5.99</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.3359</td>\n",
       "      <td>7.37</td>\n",
       "      <td>1799</td>\n",
       "      <td>3.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6298</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_12149</td>\n",
       "      <td>59357</td>\n",
       "      <td>23686</td>\n",
       "      <td>334.53</td>\n",
       "      <td>334.53</td>\n",
       "      <td>821.66</td>\n",
       "      <td>93.36</td>\n",
       "      <td>0.3261</td>\n",
       "      <td>83.21</td>\n",
       "      <td>49392</td>\n",
       "      <td>108.36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9328</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_3325</td>\n",
       "      <td>29397</td>\n",
       "      <td>3</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.3057</td>\n",
       "      <td>1.07</td>\n",
       "      <td>316</td>\n",
       "      <td>0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9329</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_3469</td>\n",
       "      <td>56074</td>\n",
       "      <td>2448</td>\n",
       "      <td>23.24</td>\n",
       "      <td>23.24</td>\n",
       "      <td>48.45</td>\n",
       "      <td>12.08</td>\n",
       "      <td>0.3430</td>\n",
       "      <td>77.75</td>\n",
       "      <td>43600</td>\n",
       "      <td>15.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9330</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_7091</td>\n",
       "      <td>66998</td>\n",
       "      <td>2</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.3245</td>\n",
       "      <td>0.57</td>\n",
       "      <td>380</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9331</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_7377</td>\n",
       "      <td>35885</td>\n",
       "      <td>2</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.3317</td>\n",
       "      <td>1.34</td>\n",
       "      <td>480</td>\n",
       "      <td>0.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9332</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_8225</td>\n",
       "      <td>178222</td>\n",
       "      <td>3</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.3687</td>\n",
       "      <td>0.36</td>\n",
       "      <td>646</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2159 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample        contig  length  Reads    RPKM    FPKM     SPM  \\\n",
       "6294   33855  contig_10982   14547   1228   70.77   70.77  173.98   \n",
       "6295   33855  contig_11000   18137    350   16.18   16.18   38.78   \n",
       "6296   33855  contig_11002  106707    894    7.02    7.02   17.31   \n",
       "6297   33855  contig_11067   24409     72    2.47    2.47    5.99   \n",
       "6298   33855  contig_12149   59357  23686  334.53  334.53  821.66   \n",
       "...      ...           ...     ...    ...     ...     ...     ...   \n",
       "9328   34044   contig_3325   29397      3    0.05    0.04    0.11   \n",
       "9329   34044   contig_3469   56074   2448   23.24   23.24   48.45   \n",
       "9330   34044   contig_7091   66998      2    0.02    0.02    0.04   \n",
       "9331   34044   contig_7377   35885      2    0.03    0.03    0.08   \n",
       "9332   34044   contig_8225  178222      3    0.01    0.01    0.02   \n",
       "\n",
       "      averageFold  referenceGC  coveragePercentage  coverageBases  medianFold  \n",
       "6294        19.77       0.3652               88.46          12868       23.69  \n",
       "6295         4.41       0.3337               48.42           8782        7.67  \n",
       "6296         1.97       0.3259               23.40          24971        6.93  \n",
       "6297         0.68       0.3359                7.37           1799        3.76  \n",
       "6298        93.36       0.3261               83.21          49392      108.36  \n",
       "...           ...          ...                 ...            ...         ...  \n",
       "9328         0.03       0.3057                1.07            316        0.27  \n",
       "9329        12.08       0.3430               77.75          43600       15.53  \n",
       "9330         0.01       0.3245                0.57            380        0.13  \n",
       "9331         0.02       0.3317                1.34            480        0.18  \n",
       "9332         0.01       0.3687                0.36            646        0.09  \n",
       "\n",
       "[2159 rows x 12 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ourcontigs = contigs[contigs['sample'].isin(wehave)]\n",
    "ourcontigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>contig</th>\n",
       "      <th>Reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6294</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>1228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6295</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_11000</td>\n",
       "      <td>350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6296</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_11002</td>\n",
       "      <td>894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6297</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_11067</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6298</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_12149</td>\n",
       "      <td>23686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9328</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_3325</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9329</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_3469</td>\n",
       "      <td>2448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9330</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_7091</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9331</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_7377</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9332</th>\n",
       "      <td>34044</td>\n",
       "      <td>contig_8225</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2159 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample        contig  Reads\n",
       "6294   33855  contig_10982   1228\n",
       "6295   33855  contig_11000    350\n",
       "6296   33855  contig_11002    894\n",
       "6297   33855  contig_11067     72\n",
       "6298   33855  contig_12149  23686\n",
       "...      ...           ...    ...\n",
       "9328   34044   contig_3325      3\n",
       "9329   34044   contig_3469   2448\n",
       "9330   34044   contig_7091      2\n",
       "9331   34044   contig_7377      2\n",
       "9332   34044   contig_8225      3\n",
       "\n",
       "[2159 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ourcontshort = ourcontigs[['sample', 'contig', 'reads']]\n",
    "ourcontshort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>contig</th>\n",
       "      <th>Reads</th>\n",
       "      <th>length</th>\n",
       "      <th>ref</th>\n",
       "      <th>coding</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>33855</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>1228</td>\n",
       "      <td>14547</td>\n",
       "      <td>Kehishuvirus</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>33861</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>48</td>\n",
       "      <td>14547</td>\n",
       "      <td>Kehishuvirus</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33898</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>2</td>\n",
       "      <td>14547</td>\n",
       "      <td>Kehishuvirus</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>33908</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>1</td>\n",
       "      <td>14547</td>\n",
       "      <td>Kehishuvirus</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>33940</td>\n",
       "      <td>contig_10982</td>\n",
       "      <td>4</td>\n",
       "      <td>14547</td>\n",
       "      <td>Kehishuvirus</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154</th>\n",
       "      <td>33992</td>\n",
       "      <td>contig_11554</td>\n",
       "      <td>437</td>\n",
       "      <td>30342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155</th>\n",
       "      <td>33993</td>\n",
       "      <td>contig_11554</td>\n",
       "      <td>2</td>\n",
       "      <td>30342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156</th>\n",
       "      <td>34000</td>\n",
       "      <td>contig_242504</td>\n",
       "      <td>319</td>\n",
       "      <td>15858</td>\n",
       "      <td>Drivevirus gastrointestinalis</td>\n",
       "      <td>11</td>\n",
       "      <td>Intestiviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2157</th>\n",
       "      <td>34001</td>\n",
       "      <td>contig_242504</td>\n",
       "      <td>2</td>\n",
       "      <td>15858</td>\n",
       "      <td>Drivevirus gastrointestinalis</td>\n",
       "      <td>11</td>\n",
       "      <td>Intestiviridae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2158</th>\n",
       "      <td>34011</td>\n",
       "      <td>contig_11676</td>\n",
       "      <td>8</td>\n",
       "      <td>36325</td>\n",
       "      <td>Mahstovirus</td>\n",
       "      <td>11</td>\n",
       "      <td>Steigviridae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2159 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample         contig  Reads  length                            ref  \\\n",
       "0      33855   contig_10982   1228   14547                   Kehishuvirus   \n",
       "1      33861   contig_10982     48   14547                   Kehishuvirus   \n",
       "2      33898   contig_10982      2   14547                   Kehishuvirus   \n",
       "3      33908   contig_10982      1   14547                   Kehishuvirus   \n",
       "4      33940   contig_10982      4   14547                   Kehishuvirus   \n",
       "...      ...            ...    ...     ...                            ...   \n",
       "2154   33992   contig_11554    437   30342                            NaN   \n",
       "2155   33993   contig_11554      2   30342                            NaN   \n",
       "2156   34000  contig_242504    319   15858  Drivevirus gastrointestinalis   \n",
       "2157   34001  contig_242504      2   15858  Drivevirus gastrointestinalis   \n",
       "2158   34011   contig_11676      8   36325                    Mahstovirus   \n",
       "\n",
       "     coding          family  \n",
       "0        11    Steigviridae  \n",
       "1        11    Steigviridae  \n",
       "2        11    Steigviridae  \n",
       "3        11    Steigviridae  \n",
       "4        11    Steigviridae  \n",
       "...     ...             ...  \n",
       "2154     11         unknown  \n",
       "2155     11         unknown  \n",
       "2156     11  Intestiviridae  \n",
       "2157     11  Intestiviridae  \n",
       "2158     11    Steigviridae  \n",
       "\n",
       "[2159 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annocont = pd.merge(ourcontshort, crassshort, on='contig')\n",
    "annocont"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "annocont.to_csv(\"our_crassphage.tsv.gz\", sep=\"\\t\", compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>contig</th>\n",
       "      <th>length</th>\n",
       "      <th>reads</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>SPM</th>\n",
       "      <th>averageFold</th>\n",
       "      <th>referenceGC</th>\n",
       "      <th>coveragePercentage</th>\n",
       "      <th>coverageBases</th>\n",
       "      <th>MedianFold</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_102265</td>\n",
       "      <td>1524</td>\n",
       "      <td>4</td>\n",
       "      <td>1.38</td>\n",
       "      <td>1.38</td>\n",
       "      <td>5.14</td>\n",
       "      <td>0.81</td>\n",
       "      <td>0.4337</td>\n",
       "      <td>26.57</td>\n",
       "      <td>405</td>\n",
       "      <td>1.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_115141</td>\n",
       "      <td>1423</td>\n",
       "      <td>3</td>\n",
       "      <td>1.11</td>\n",
       "      <td>0.74</td>\n",
       "      <td>3.74</td>\n",
       "      <td>0.59</td>\n",
       "      <td>0.4329</td>\n",
       "      <td>42.16</td>\n",
       "      <td>600</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_147529</td>\n",
       "      <td>1247</td>\n",
       "      <td>4</td>\n",
       "      <td>1.68</td>\n",
       "      <td>1.68</td>\n",
       "      <td>6.54</td>\n",
       "      <td>1.02</td>\n",
       "      <td>0.4338</td>\n",
       "      <td>37.37</td>\n",
       "      <td>466</td>\n",
       "      <td>1.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_158706</td>\n",
       "      <td>1079</td>\n",
       "      <td>64</td>\n",
       "      <td>31.15</td>\n",
       "      <td>31.15</td>\n",
       "      <td>125.67</td>\n",
       "      <td>19.70</td>\n",
       "      <td>0.4217</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1079</td>\n",
       "      <td>11.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7544</td>\n",
       "      <td>contig_217056</td>\n",
       "      <td>1035</td>\n",
       "      <td>6</td>\n",
       "      <td>3.04</td>\n",
       "      <td>3.04</td>\n",
       "      <td>9.17</td>\n",
       "      <td>1.44</td>\n",
       "      <td>0.4415</td>\n",
       "      <td>47.92</td>\n",
       "      <td>496</td>\n",
       "      <td>1.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>38209</td>\n",
       "      <td>contig_58328</td>\n",
       "      <td>2220</td>\n",
       "      <td>25</td>\n",
       "      <td>8.09</td>\n",
       "      <td>7.77</td>\n",
       "      <td>25.07</td>\n",
       "      <td>3.48</td>\n",
       "      <td>0.4667</td>\n",
       "      <td>20.18</td>\n",
       "      <td>448</td>\n",
       "      <td>7.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>38210</td>\n",
       "      <td>contig_58328</td>\n",
       "      <td>2220</td>\n",
       "      <td>179</td>\n",
       "      <td>29.33</td>\n",
       "      <td>29.17</td>\n",
       "      <td>34.11</td>\n",
       "      <td>20.49</td>\n",
       "      <td>0.4667</td>\n",
       "      <td>27.75</td>\n",
       "      <td>616</td>\n",
       "      <td>42.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_121925</td>\n",
       "      <td>1378</td>\n",
       "      <td>2</td>\n",
       "      <td>0.69</td>\n",
       "      <td>0.69</td>\n",
       "      <td>2.55</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.4862</td>\n",
       "      <td>33.45</td>\n",
       "      <td>461</td>\n",
       "      <td>0.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_23429</td>\n",
       "      <td>8205</td>\n",
       "      <td>2</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.45</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.4924</td>\n",
       "      <td>6.17</td>\n",
       "      <td>506</td>\n",
       "      <td>0.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>38211</td>\n",
       "      <td>contig_58328</td>\n",
       "      <td>2220</td>\n",
       "      <td>1</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.4667</td>\n",
       "      <td>10.54</td>\n",
       "      <td>234</td>\n",
       "      <td>0.31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>799 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample         contig  length  reads   RPKM   FPKM     SPM  averageFold  \\\n",
       "0      7544  contig_102265    1524      4   1.38   1.38    5.14         0.81   \n",
       "1      7544  contig_115141    1423      3   1.11   0.74    3.74         0.59   \n",
       "2      7544  contig_147529    1247      4   1.68   1.68    6.54         1.02   \n",
       "3      7544  contig_158706    1079     64  31.15  31.15  125.67        19.70   \n",
       "4      7544  contig_217056    1035      6   3.04   3.04    9.17         1.44   \n",
       "..      ...            ...     ...    ...    ...    ...     ...          ...   \n",
       "794   38209   contig_58328    2220     25   8.09   7.77   25.07         3.48   \n",
       "795   38210   contig_58328    2220    179  29.33  29.17   34.11        20.49   \n",
       "796   38211  contig_121925    1378      2   0.69   0.69    2.55         0.50   \n",
       "797   38211   contig_23429    8205      2   0.12   0.12    0.45         0.09   \n",
       "798   38211   contig_58328    2220      1   0.22   0.00    0.53         0.11   \n",
       "\n",
       "     referenceGC  coveragePercentage  coverageBases  MedianFold  \n",
       "0         0.4337               26.57            405        1.46  \n",
       "1         0.4329               42.16            600        0.76  \n",
       "2         0.4338               37.37            466        1.35  \n",
       "3         0.4217              100.00           1079       11.30  \n",
       "4         0.4415               47.92            496        1.62  \n",
       "..           ...                 ...            ...         ...  \n",
       "794       0.4667               20.18            448        7.29  \n",
       "795       0.4667               27.75            616       42.13  \n",
       "796       0.4862               33.45            461        0.77  \n",
       "797       0.4924                6.17            506        0.37  \n",
       "798       0.4667               10.54            234        0.31  \n",
       "\n",
       "[799 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rdrp = pd.read_csv('rdrp_contig_count_table.tsv.gz',  sep=\"\\t\", compression='gzip')\n",
    "rdrp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "rdrpshort = rdrp[['sample', 'contig', 'reads']]\n",
    "ourrdrp = rdrpshort[rdrpshort['sample'].isin(wehave)]\n",
    "ourrdrp.to_csv(\"rdrp_wehave.tsv.gz\",  sep=\"\\t\", compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
